tests/lib/linter/file-report.js

```diff
@@ -0,0 +1,1441 @@
+/**
+ * @fileoverview Tests for FileReport class
+ * @author ESLint
+ */
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+const assert = require("chai").assert;
+const { FileReport } = require("../../../lib/linter/file-report");
+const { SourceCode } = require("../../../lib/languages/js/source-code");
+const espree = require("espree");
+
+//------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+/**
+ * Creates a SourceCode instance out of JavaScript text
+ * @param {string} text Source text
+ * @returns {SourceCode} A SourceCode instance for that text
+ */
+function createSourceCode(text) {
+	return new SourceCode(
+		text,
+		espree.parse(text.replace(/^\uFEFF/u, ""), {
+			loc: true,
+			range: true,
+			raw: true,
+			tokens: true,
+			comment: true,
+		}),
+	);
+}
+
+/**
+ * Mock rule mapper for testing
+ * @returns {Object} Mock rule definition
+ */
+function mockRuleMapper() {
+	return {
+		meta: {
+			messages: {
+				testMessage: "foo",
+				suggestion1: "First suggestion",
+				suggestion2: "Second suggestion {{interpolated}}",
+			},
+		},
+	};
+}
+
+const language = { columnStart: 0, lineStart: 1 };
+
+//------------------------------------------------------------------------------
+// Tests
+//------------------------------------------------------------------------------
+
+describe("FileReport", () => {
+	let sourceCode, node, location, fileReport;
+
+	beforeEach(() => {
+		sourceCode = createSourceCode("foo\nbar");
+		node = sourceCode.ast.body[0];
+		location = sourceCode.ast.body[1].loc.start;
+		fileReport = new FileReport({
+			ruleMapper: mockRuleMapper,
+			sourceCode,
+			language,
+		});
+	});
+
+	describe("addRuleMessage", () => {
+		it("should add a message with a string message", () => {
+			const result = fileReport.addRuleMessage(
+				"foo-rule",
+				2,
+				node,
+				location,
+				"foo",
+				{},
+			);
+
+			assert.deepStrictEqual(result, {
+				ruleId: "foo-rule",
+				severity: 2,
+				message: "foo",
+				line: 2,
+				column: 1,
+				nodeType: "ExpressionStatement",
+			});
+			assert.strictEqual(fileReport.messages.length, 1);
+			assert.deepStrictEqual(fileReport.messages[0], result);
```

Comment:
Perhaps these checks should use `strictEqual()`? I'm guessing the purpose of returning created lint problems is that the caller can additionally modify them if needed, for which it would be important that the same instance is stored in `FileReport#messages`.